botocore/utils.py

```diff
@@ -274,29 +281,56 @@ def __init__(self, timeout=DEFAULT_METADATA_SERVICE_TIMEOUT,
             proxies=get_environ_proxies(self._base_url),
         )
 
-    def _get_request(self, url_path, retry_func):
+    def _fetch_metadata_token(self):
+        self._assert_enabled()
+        url = self._base_url + self._TOKEN_PATH
+        headers = {
+            'x-aws-ec2-metadata-token-ttl-seconds': self._TOKEN_TTL,
+        }
+        self._add_user_agent(headers)
+        request = botocore.awsrequest.AWSRequest(
+            method='PUT', url=url, headers=headers)
+        for i in range(self._num_attempts):
+            try:
+                response = self._session.send(request.prepare())
+                if response.status_code == 200:
+                    return response.text
+                elif response.status_code in (404, 403):
+                    return None
+                elif response.status_code in (400,):
+                    raise BadIMDSRequestError(request)
+            except ReadTimeoutError:
```

Comment:
Are we sure that this will always get raised as a read timeout? When I was working on the `urllib3` changes I remember some scenarios where it was ambiguous exactly what timeout error would get raised.

Reply:
In the repro case it appears to only be that error.